### PR TITLE
Buttons: Preserve styling from adjacent button blocks when inserting a new button block

### LINKED ISSUE #37904: Buttons Block: Clicking inserter icon to add a button should use styling of an existing button in the list
## What problem does this address?

When you hit "enter" from within a Button block, the added button block uses the same styling as the existing button block, which makes it nice and easy to edit a list of similarly styled buttons. However, the inserter button displayed at the end of the Buttons block currently doesn't respect the styling of the existing buttons.

## What is your proposed solution?

When you click the + icon at the end of a list of Button blocks to insert an additional Button, let's see about grabbing the styling / attributes of the last Button block in the list and applying that to the inserted block, consistent with the Enter key behaviour.

Here is a screengrab of the current behaviour when we hit "enter" to add a Button block, and when we click the inserter at the bottom of the Buttons block:

https://user-images.githubusercontent.com/14988353/149066785-023c8a17-43bb-464f-93b6-3da133002548.mp4

### PR DESCRIPTION
**Pull request by andrewserong**
## Description

Fixes #37904

This PR updates the Inserter logic for the `defaultBlock` to attempt to use attributes from an adjacent similar block when inserting a new block directly between two blocks. The use case is primarily for the Buttons block, where clicking between two individual Button blocks inserts a new button. For layouts or patterns that use full-width buttons (e.g. a Link in Bio pattern), or buttons with particular styling, it might make it smoother for users if in the inserted button we use those existing attributes.

I'd love feedback on whether or not this is universal enough to be added to the `defaultBlock` logic added by @tellthemachines in #34899.

The logic in this PR includes:

* In the inserter, when a block is being directly inserted via the inserter, attempt to get the attributes from the previous block if it is of the same block type, and the `directInsertBlock` has an `attributesToCopy` array with matching attribute names.
* If that isn't possible (e.g. you're clicking the last inserter in the Buttons block) then attempt to get the attributes from the last of the inner blocks.
* I've refactored the `directInsertBlock` to be an object instead of an array (to allow for the `attributesToCopy` prop) and added a typedef

For a precedent surrounding the suggested behaviour when inserting the button block, currently if you hit "enter" on a button block to fire the `onSplit` function, it [uses the existing attributes](https://github.com/wordpress/gutenberg/blob/8f4726f339332f4008e32e9906c93ee72a1f791b/packages/block-library/src/button/edit.js#L186) when creating the new button block.

## How has this been tested?

1. Insert a Buttons block and add a few buttons. Set them to 100% width and / or some color styles.
2. Hover between the blocks and click the inserter to add a Button block. The styling of the previous Button block should be preserved.
3. Select the parent Buttons block and click the inserter at the end of the list of Button blocks. The newly inserted Button block should preserve the styling/attributes of the last block in the list.
4. Ensure that link / text is not preserved in the inserted block.
5. Double-check that this change doesn't introduce regressions for those blocks that already use the `__experimentalDefaultBlock` option when registering the block (e.g. Navigation and Navigation Submenu blocks)

## Screenshots 

https://user-images.githubusercontent.com/14988353/149276720-6f213efb-4c8d-4c8f-899c-d3308c4f265d.mp4

## Types of changes

Somewhere between a bug fix and an enhancement (I'd argue it's an enhancement but it will likely feel like a bug fix to some users)

## Checklist:
- [ ] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). 
- [ ] I've updated related schemas if appropriate.

### CODE REVIEW

**Comment by andrewserong** [comment]
@tellthemachines I've added you as a reviewer since you worked on adding the `defaultBlock` behaviour to the inserter and have more experience with the Navigation block. This isn't at all urgent, though, so no worries if you don't have time to review!

**Review by stacimc — CHANGES_REQUESTED** [review]
This would be such a welcome change! IMO it dramatically improves the experience of working with the Buttons block. I'm interested to hear what others think, but I also like its addition to the `defaultBlock` logic and its potential to be used more universally. I appreciate there's already handling for inner blocks of different types :)

This tests great for me with the Buttons block! I've tried adding all possible attributes, and inserting buttons in different positions. Attributes are correctly taken from the previous block, and text/url are not copied. This feels so much more intuitive ✨ 

https://user-images.githubusercontent.com/63313398/149397648-e108d5a2-5221-4568-9a48-d2e7ab8ababd.mov

I just caught one regression with the Navigation block. When you add a submenu and attempt to insert links, you'll get the inserter on the first link. If you try to add additional links, though, it just copies the first link over rather than opening the inserter. I think you'll just need to mark some link attributes like `url` as undefined so they don't get copied over -- looks like maybe in the [navigation-submenu DEFAULT_BLOCK](https://github.com/WordPress/gutenberg/blob/94d553092160adc2e1f9d9362ee87c158a67fea8/packages/block-library/src/navigation-submenu/edit.js#L53)?

https://user-images.githubusercontent.com/63313398/149398180-2314cb77-f0f6-483e-9103-88010a3c1964.mov

Other than that, Navigation also looks good to me!

**Comment by andrewserong** [comment]
Thanks for reviewing @stacimc! Good catch with the Navigation and Navigation Submenu blocks — I _think_ from the looks of it the Navigation Link block mostly has attributes that we wouldn't want to copy over when inserting a block. However, I quite like the default behaviour of copying across attributes, so in https://github.com/WordPress/gutenberg/pull/37905/commits/29774c9a26ec9a4cc718833b9f7269d041a30790 I've had a go at setting the Navigation Link's `DEFAULT_BLOCK` to use all default or undefined attributes. It works, but it might not be the tidiest way to do it, particularly if it's more common that blocks will have lots of attributes that shouldn't be copied over 🤔

Happy for any feedback if folks think there's a better way to handle it (particularly from folks who've done more work on the Navigation block) — I was trying to avoid adding another flag to switch on the behaviour in this PR, but there might be a trade-off to be found if the approach of setting defaults is too unwieldy!

Anyway, here's a screengrab of the navigation link insertion after the above commit was pushed (I think it's working properly now):

https://user-images.githubusercontent.com/14988353/149434584-da12b8a3-58e7-43eb-a1c4-565aa1ee8675.mp4

File: packages/block-editor/src/components/inserter/index.js
Comment (tellthemachines) [inline packages/block-editor/src/components/inserter/index.js]: This ternary is a little long for comfortable reading; could we turn it into an if else? Perhaps we can conflate it with the ternary below and just have something like
```
let blockToInsert
if ( directInsertBlock?.length ) 
   blockToInsert =  createBlock( // get attribs etc )
else
  blockToInsert =  createBlock( allowedBlockType.name )
```

File: packages/block-library/src/navigation-submenu/edit.js
Comment (tellthemachines) [inline packages/block-library/src/navigation-submenu/edit.js]: It's not great that we have to add exceptions to copied attributes in every single place we use direct insert. We may not be using it in a lot of core blocks (and there probably won't be many more), but if/when it becomes stable, we want third party block developers to find it easy to use.

One alternative would be to specify in DEFAULT_BLOCK a list of attributes we want copied, and have `getAdjacentBlockAttributes` only return those; then by default if nothing is specified nothing is copied. 

Or we could add an `__experimentalCopyStyleAttributes` instead of leveraging direct insert, but what we're doing here is close enough to direct insert that it kind of makes sense to have it there.

**Review by tellthemachines — COMMENTED** [review]
Thanks for addressing this issue! I left a few comments below; essentially I think we should make it so existing instances of `__experimentalDefaultBlock` don't have to be changed at all.

File: packages/block-library/src/navigation-submenu/edit.js
Comment (andrewserong) [reply packages/block-library/src/navigation-submenu/edit.js]: > It's not great that we have to add exceptions to copied attributes in every single place we use direct insert.

Definitely not, thanks for taking a look here!

> One alternative would be to specify in DEFAULT_BLOCK a list of attributes we want copied, and have getAdjacentBlockAttributes only return those; then by default if nothing is specified nothing is copied.

Oh, I like that idea — I felt a bit uneasy just copying everything, and I think it makes it cleaner if we can keep it a part of `DEFAULT_BLOCK` if possible without adding an extra flag.

At the moment `DEFAULT_BLOCK` is an array so that it matches the arguments we pass to `createBlock`. I was wondering if it'd make sense to switch it to an object, so that we could add an extra key for the attributes to be copied, and have something like the following?

```
const DEFAULT_BLOCK = {
  name: 'core/my-block',
  attributes: {
    myDefaultAttribute: 'default value to set'
  },
  attributesToCopy: {
    myCopiedAttribute: true
  }
}
```

I'll keep chipping away at this PR next week, but let me know if you have a particular preference for how we implement it. Either way, explicitly flagging when we'd like to copy attributes sounds safer than assuming and having block authors have to add exceptions 🙂

File: packages/block-editor/src/components/inserter/index.js
Comment (andrewserong) [reply packages/block-editor/src/components/inserter/index.js]: Thanks that's a good suggestion, I'll have a play with that. I struggled a little with this PR in trying to keep the logic neat (I mostly hacked it together and then tried in vain to get it presentable 😅), so I very much appreciate the extra set of 👀!

**Comment by andrewserong** [comment]
@tellthemachines and @stacimc, I've had a go at refactoring this so that it's opt-in to copy over attributes from adjacent blocks rather than opt-out, so that 🤞 it's a little more straightforward for either existing blocks, or 3rd party block authors to work with (i.e. they don't need to know to exclude attributes from being copied).

To get there, I've also refactored usage of the `directInsertBlock` to be an object with props `name`, and optional props `attributes` and `attributesToCopy` — I _think_ this makes it a little cleaner than the array that's tied to the signature of `createBlock`, but do let me know if you'd prefer we don't change this! For clarity over its usage, I've added a `@typedef` for the `WPDirectInsertBlock` to the `__experimentalGetDirectInsertBlock` selector.

Based on [a comment](https://github.com/WordPress/gutenberg/pull/34899/files#r71870893) from @gwwar on the earlier PR that introduced the `directInsertBlock` I've left off `innerBlocks` from the type and the call to `createBlock`, but that should be easy enough to add back in if folks prefer we keep it, too. Apologies if I've missed some context there!

I'm pretty sure this PR is now ready for another review, but it's the end of my day now so I haven't tested it fully just yet. Will do some more testing tomorrow!

**Comment by andrewserong** [comment]
Alrighty, I _think_ this PR is working correctly now (fixed a regression with the Navigation block inserter) 👍

**Review by stacimc — COMMENTED** [review]
This is looking good to me! Copying nothing by default and requiring configuration of attributes to copy in the `WPDirectInsertBlock` seems very reasonable; it is more intentional and probably less likely to cause error if misconfigured.

I'll defer to those more familiar with the Navigation block and the `directInsertBlock` implementation, but the use of the object looks okay to me with the addition of the typedef.   If I read the [linked conversation](https://github.com/WordPress/gutenberg/pull/34899/files#r718708930) correctly, there's no current use case for including the `innerBlocks` and it's safer to omit it for now, so I would hazard you've got the right idea there as well.

<hr/>

- [x] Style attributes are copied from the previous button block
- [x] Url & text attributes are not copied
- [x] Works when inserting between two existing buttons, or at the end of the list

I re-tested Navigation as well:

- [x] Links are added by default
- [x] No attributes are copied over as new links are added
- [x] Submenus can be added, and links are added to them by default

As soon as you add a non-link block to a Navigation menu (for example, if you add the Site Logo), clicking `Insert` will bring up the full block inserter menu. This is what's happening on trunk as well, and I believe that's the intended behavior.

File: packages/block-editor/src/components/inserter/index.js
Comment (stacimc) [inline packages/block-editor/src/components/inserter/index.js:356]: Just for my curiosity and to see if I missed a test case, did you have a use case in mind where the `directInsertBlock` would override attributes of the adjacent block?

File: packages/block-editor/src/components/inserter/index.js
Comment (tellthemachines) [inline packages/block-editor/src/components/inserter/index.js:353]: Hmm, this will cycle through all the adjacent block attributes even if we don't mean to copy any of them. I'm thinking it would be better to return early if there are no attributes to be copied. A couple ways we could do this:

* Wrap this logic inside another if statement checking for `directInsertBlock.attributesToCopy`; or
* pass the attributes to copy to `getAdjacentBlockAttributes` and have it return only those attributes; in that case we can return early from that function if nothing is passed to it. I'm more inclined towards this option as it would tidy all the extra attribute logic inside `getAdjacentBlockAttributes`. What do you think?

File: packages/block-editor/src/components/inserter/index.js
Comment (tellthemachines) [reply packages/block-editor/src/components/inserter/index.js:356]: Good question. `directInsertBlock` could have attributes of its own defined; in case we define an attribute and also want to grab the same attribute from an adjacent block, it might make more sense for the adjacent one to override the initial.

File: packages/block-library/src/buttons/edit.js
Comment (tellthemachines) [inline packages/block-library/src/buttons/edit.js]: Do we also want to copy border and spacing styles?

**Review by tellthemachines — COMMENTED** [review]
Thanks for the quick iteration! It's much improved; just a few more comments below.

File: packages/block-library/src/buttons/edit.js
Comment (andrewserong) [reply packages/block-library/src/buttons/edit.js]: Oh, yes, and font settings, too! I totally forgot about those attributes provided by block supports. I've added a few more here and it seems to be working well:

Kapture 2022-01-19 at 16 52 11

File: packages/block-editor/src/components/inserter/index.js
Comment (andrewserong) [reply packages/block-editor/src/components/inserter/index.js:356]: Agreed, thanks for asking the question! I wasn't _too_ sure of a use case, but thought it'd still be good to allow `attributes` to be defined in the `directInsertBlock`. Allowing the adjacent ones to override sounds good to me, I've swapped the order here 👍

File: packages/block-editor/src/components/inserter/index.js
Comment (andrewserong) [reply packages/block-editor/src/components/inserter/index.js:353]: Great points! I've moved the logic up to go inside `getAdjacentBlockAttributes`. That function is a _little_ long now, but it feels a bit cleaner putting it there, and also only iterating over the specified `attributesToCopy` if provided rather than every attribute of the adjacent block 👍 

Let me know if you think it needs further tidying (or refactoring to another file now that it's a bit longer!)

**Comment by andrewserong** [comment]
Thanks again for the reviews @stacimc and @tellthemachines! I _think_ I've implemented all the feedback, but let me know if I've missed anything 😀

File: packages/block-editor/src/components/inserter/index.js
Comment (stacimc) [inline packages/block-editor/src/components/inserter/index.js]: Can we remove this and just check:

`if ( adjacentAttributes.hasOwnProperty( attribute ) )`?

**Review by stacimc — COMMENTED** [review]
+1 for moving the logic into `getAdjacentAttributes`, which also has the advantage of making `getInsertionIndex` more readable.

I neglected to test all attributes on my second run-through, sorry! Tested again and all style attributes are being copied over, including attrs from block supports and additional CSS classes 😄

Just one question, but LGTM!

File: packages/block-editor/src/components/inserter/index.js
Comment (andrewserong) [reply packages/block-editor/src/components/inserter/index.js]: Thanks for re-testing @stacimc, another good question! I wasn't too sure about this one — the `attributesToCopy` is currently an object where each attribute is set to a boolean (`true` in all the cases in the Buttons block). So, if someone were to theoretically set one of them to `false` it might be odd if it still copied the attribute.

I initially had it an object so that look up was easy when looping over the adjacent block's attributes, so an object look up was preferable to an array. Now that it's switched the other way (we're looping over the `attributesToCopy`) the object structure doesn't seem to be _quite_ as necessary. I kinda like the parallel, though, with having both `attributes` and `attributesToCopy` being objects.

So, a couple of follow-up questions! 😀

Is it better for `attributesToCopy` to be an object or an array? And if we keep it as an object, should it be possible for someone to set one of the values to `false` to switch off copying of that attribute (there's no real use case for this, but I _think_ it determines whether or not we should remove this line 🤔)

Apologies if I'm overthinking this!

Comment (stacimc) [reply packages/block-editor/src/components/inserter/index.js]: Oh of course, `attributesToCopy` is an object now 🤦‍♀️  You're right, this is a bit tricky! Provided `attributesToCopy` remains an object, I'd leave it as you have it. I don't think anyone _ought_ to explicitly set a value to `false` within that object, but I think you're right to safeguard against it.

To me it does feel slightly less clear when reading -- maybe just a naming thing, but I wouldn't expect a parameter named `attributesToCopy` to contain data for attributes that _shouldn't_ be copied. I may be the one overthinking it here, though 😄 I don't feel strongly about changing it and I don't think it should block the PR, so up to you!

**Review by stacimc — APPROVED** [review]
This is testing well and makes a big impact on the usability of the Buttons block 🥳 Very excited to see this one go in!

File: packages/block-editor/src/components/inserter/index.js
Comment (tellthemachines) [reply packages/block-editor/src/components/inserter/index.js]: Making `attributesToCopy` an array would make it less verbose to add those attributes, and semantically makes more sense because we wouldn't ever expect them to be set to `false` (there's no point adding an attribute to copy if we don't want it copied, and there's no way of toggling those attributes after they're set). 

I agree it's not a blocking issue for this PR though; we can always address it in a follow up.

**Review by tellthemachines — APPROVED** [review]
This is looking good now 🎉

File: packages/block-editor/src/components/inserter/index.js
Comment (andrewserong) [reply packages/block-editor/src/components/inserter/index.js]: Thanks @tellthemachines, that's a great point. I've updated it to be an array in https://github.com/WordPress/gutenberg/pull/37905/commits/401358fc268b6cc74033650a3b203adb032b52b8 (and updated the `typedef` so that it's an array of strings), and that looks much clearer to me. Cheers!

I'm out for the next few hours, but will 🤞 merge in this PR tonight once the tests pass, unless there are any objections 🙂

**Pull request merged**
